From: David Curtiss <david.curtiss@ni.com>
Subject: [PATCH] ws2_32: Allow getsockname after AcceptEx
Message-Id: <20220407160232.244307-1-david.curtiss@ni.com>
Date: Thu, 7 Apr 2022 09:02:32 -0700

.NET 6's HTTP/Socket code queries this. Winsock allows getsockname
on the AcceptEx AcceptSocket, but only if SO_UPDATE_ACCEPT_CONTEXT
is set.

Signed-off-by: David Curtiss <david.curtiss@ni.com>
---
 dlls/ws2_32/tests/sock.c | 9 +++++++++
 server/sock.c            | 2 +-
 2 files changed, 10 insertions(+), 1 deletion(-)

diff --git a/dlls/ws2_32/tests/sock.c b/dlls/ws2_32/tests/sock.c
index 11111111111..11111111111 100644
--- a/dlls/ws2_32/tests/sock.c
+++ b/dlls/ws2_32/tests/sock.c
@@ -7931,6 +7931,15 @@ static void test_AcceptEx(void)
     ok(bret, "GetOverlappedResult failed, error %ld\n", GetLastError());
     ok(bytesReturned == 0, "bytesReturned isn't supposed to be %ld\n", bytesReturned);
 
+    /* Try to call getsockname on the acceptor socket.
+     *
+     * On Windows, this requires setting SO_UPDATE_ACCEPT_CONTEXT. */
+    iret = setsockopt(acceptor, SOL_SOCKET, SO_UPDATE_ACCEPT_CONTEXT, (char *)&listener, sizeof(SOCKET));
+    ok(!iret, "Failed to set accept context %ld\n", GetLastError());
+    iret = getsockname(acceptor, (struct sockaddr *)&peerAddress, &remoteSize);
+    ok(!iret, "getsockname failed.\n");
+    ok(remoteSize == sizeof(struct sockaddr_in), "got remote size %u\n", remoteSize);
+
     closesocket(connector);
     connector = INVALID_SOCKET;
     closesocket(acceptor);
diff --git a/server/sock.c b/server/sock.c
index 11111111111..11111111111 100644
--- a/server/sock.c
+++ b/server/sock.c
@@ -2739,7 +2739,7 @@ static void sock_ioctl( struct fd *fd, ioctl_code_t code, struct async *async )
     }
 
     case IOCTL_AFD_GETSOCKNAME:
-        if (!sock->bound)
+        if (!sock->addr_len)
         {
             set_error( STATUS_INVALID_PARAMETER );
             return;

-- 
2.36.1

